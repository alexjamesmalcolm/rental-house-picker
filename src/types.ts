export interface Listing {
  buildings: Building[];
  price: number;
}

export interface ListingArrangement {
  buildingArrangements: BuildingArrangement[];
  listing: Listing;
}

export interface Building {
  floors: Floor[];
}

export interface BuildingArrangement {
  floorArrangements: FloorArrangement[];
}

export interface Floor {
  rooms: Room[];
}

export interface FloorArrangement {
  roomArrangements: RoomArrangement[];
}

export interface Room {
  beds: Bed[];
}

export interface RoomArrangement {
  bedArrangements: BedArrangement[];
}

export interface KingBed {
  name: "king";
}

export interface QueenBed {
  name: "queen";
}

export interface TwinBed {
  name: "twin";
}

export type Bed = KingBed | QueenBed | TwinBed;

export interface BedArrangement {
  bed: Bed;
  people: Person[];
}

export interface Couple {
  husband: Man;
  wife: Woman;
}

export interface Family {
  couple: Couple;
  children: Person[];
}

export type Gender = "male" | "female";

export interface Man extends BasePerson {
  gender: "male";
}
export interface Woman extends BasePerson {
  gender: "female";
}
export type Person = Man | Woman;

interface BasePerson {
  name: string;
  howManyOthersCanShareBed: {
    king: number;
    queen: number;
    twin: number;
  };
}

export interface PeopleGroup {
  families: Family[];
  people: Person[];
}
