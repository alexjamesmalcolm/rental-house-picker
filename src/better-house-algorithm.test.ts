import { fillBeds, getAllPossibleArrangements } from "./better-house-algorithm";
import { Listing } from "./types";
import { homechurch } from "./homechurch";

test("SPECTACULAR NEW BEACH HOUSE WITH 4TH LEVEL SKYDECK & CUSTOM POOL/SPA", () => {
  const listings: Listing[] = [
    {
      price: 0,
      buildings: [
        {
          floors: [
            {
              rooms: [
                {
                  beds: [{ name: "king" }, { name: "twin" }, { name: "twin" }],
                },
                { beds: [{ name: "queen" }, { name: "queen" }] },
              ],
            },
            {
              rooms: [
                {
                  beds: [{ name: "king" }, { name: "twin" }, { name: "twin" }],
                },
              ],
            },
            {
              rooms: [
                {
                  beds: [{ name: "king" }, { name: "twin" }],
                },
                {
                  beds: [{ name: "king" }, { name: "twin" }],
                },
                {
                  beds: [{ name: "king" }],
                },
                {
                  beds: [
                    { name: "twin" },
                    { name: "twin" },
                    { name: "twin" },
                    { name: "twin" },
                    { name: "twin" },
                    { name: "twin" },
                  ],
                },
              ],
            },
          ],
        },
      ],
    },
  ];
  const generator = getAllPossibleArrangements({
    peopleGroup: homechurch,
    listings,
  });
  const people = generator.next().value;
  expect(fillBeds(listings, people)).toMatchSnapshot();
});
