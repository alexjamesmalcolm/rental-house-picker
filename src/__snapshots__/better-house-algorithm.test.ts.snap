// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SPECTACULAR NEW BEACH HOUSE WITH 4TH LEVEL SKYDECK & CUSTOM POOL/SPA 1`] = `
Array [
  Object {
    "buildingArrangements": Array [
      Object {
        "floorArrangements": Array [
          Object {
            "roomArrangements": Array [
              Object {
                "bedArrangements": Array [
                  Object {
                    "bed": Object {
                      "name": "king",
                    },
                    "people": Array [
                      Object {
                        "gender": "female",
                        "howManyOthersCanShareBed": Object {
                          "king": 2,
                          "queen": 1,
                          "twin": 0,
                        },
                        "name": "Sarah Branaghan",
                      },
                      Object {
                        "gender": "male",
                        "howManyOthersCanShareBed": Object {
                          "king": 2,
                          "queen": 1,
                          "twin": 0,
                        },
                        "name": "Daniel Branaghan",
                      },
                    ],
                  },
                  Object {
                    "bed": Object {
                      "name": "twin",
                    },
                    "people": Array [],
                  },
                  Object {
                    "bed": Object {
                      "name": "twin",
                    },
                    "people": Array [],
                  },
                ],
              },
              Object {
                "bedArrangements": Array [
                  Object {
                    "bed": Object {
                      "name": "queen",
                    },
                    "people": Array [
                      Object {
                        "gender": "female",
                        "howManyOthersCanShareBed": Object {
                          "king": 2,
                          "queen": 1,
                          "twin": 0,
                        },
                        "name": "Carrie Marchese",
                      },
                      Object {
                        "gender": "male",
                        "howManyOthersCanShareBed": Object {
                          "king": 2,
                          "queen": 1,
                          "twin": 0,
                        },
                        "name": "Danny Marchese",
                      },
                    ],
                  },
                  Object {
                    "bed": Object {
                      "name": "queen",
                    },
                    "people": Array [
                      Object {
                        "gender": "female",
                        "howManyOthersCanShareBed": Object {
                          "king": 2,
                          "queen": 2,
                          "twin": 1,
                        },
                        "name": "Hope",
                      },
                    ],
                  },
                ],
              },
            ],
          },
          Object {
            "roomArrangements": Array [
              Object {
                "bedArrangements": Array [
                  Object {
                    "bed": Object {
                      "name": "king",
                    },
                    "people": Array [
                      Object {
                        "gender": "female",
                        "howManyOthersCanShareBed": Object {
                          "king": 2,
                          "queen": 1,
                          "twin": 0,
                        },
                        "name": "Christa Brickey",
                      },
                      Object {
                        "gender": "female",
                        "howManyOthersCanShareBed": Object {
                          "king": 2,
                          "queen": 1,
                          "twin": 0,
                        },
                        "name": "Andrea Davis",
                      },
                      Object {
                        "gender": "female",
                        "howManyOthersCanShareBed": Object {
                          "king": 2,
                          "queen": 1,
                          "twin": 0,
                        },
                        "name": "Kaila",
                      },
                    ],
                  },
                  Object {
                    "bed": Object {
                      "name": "twin",
                    },
                    "people": Array [
                      Object {
                        "gender": "female",
                        "howManyOthersCanShareBed": Object {
                          "king": 2,
                          "queen": 1,
                          "twin": 0,
                        },
                        "name": "Serena",
                      },
                    ],
                  },
                  Object {
                    "bed": Object {
                      "name": "twin",
                    },
                    "people": Array [
                      Object {
                        "gender": "female",
                        "howManyOthersCanShareBed": Object {
                          "king": 2,
                          "queen": 1,
                          "twin": 0,
                        },
                        "name": "Sam Knight",
                      },
                    ],
                  },
                ],
              },
            ],
          },
          Object {
            "roomArrangements": Array [
              Object {
                "bedArrangements": Array [
                  Object {
                    "bed": Object {
                      "name": "king",
                    },
                    "people": Array [
                      Object {
                        "gender": "male",
                        "howManyOthersCanShareBed": Object {
                          "king": 2,
                          "queen": 1,
                          "twin": 0,
                        },
                        "name": "Alec Brickey",
                      },
                      Object {
                        "gender": "male",
                        "howManyOthersCanShareBed": Object {
                          "king": 2,
                          "queen": 1,
                          "twin": 0,
                        },
                        "name": "Travis Davis",
                      },
                      Object {
                        "gender": "male",
                        "howManyOthersCanShareBed": Object {
                          "king": 2,
                          "queen": 1,
                          "twin": 0,
                        },
                        "name": "Sicquan",
                      },
                    ],
                  },
                  Object {
                    "bed": Object {
                      "name": "twin",
                    },
                    "people": Array [
                      Object {
                        "gender": "male",
                        "howManyOthersCanShareBed": Object {
                          "king": 2,
                          "queen": 1,
                          "twin": 0,
                        },
                        "name": "Sam DeVine",
                      },
                    ],
                  },
                ],
              },
              Object {
                "bedArrangements": Array [
                  Object {
                    "bed": Object {
                      "name": "king",
                    },
                    "people": Array [
                      Object {
                        "gender": "female",
                        "howManyOthersCanShareBed": Object {
                          "king": 2,
                          "queen": 1,
                          "twin": 0,
                        },
                        "name": "Sam Donermeyer",
                      },
                      Object {
                        "gender": "female",
                        "howManyOthersCanShareBed": Object {
                          "king": 2,
                          "queen": 1,
                          "twin": 0,
                        },
                        "name": "Pearl",
                      },
                      Object {
                        "gender": "female",
                        "howManyOthersCanShareBed": Object {
                          "king": 2,
                          "queen": 1,
                          "twin": 0,
                        },
                        "name": "Molly",
                      },
                    ],
                  },
                  Object {
                    "bed": Object {
                      "name": "twin",
                    },
                    "people": Array [
                      Object {
                        "gender": "female",
                        "howManyOthersCanShareBed": Object {
                          "king": 2,
                          "queen": 1,
                          "twin": 0,
                        },
                        "name": "Mara",
                      },
                    ],
                  },
                ],
              },
              Object {
                "bedArrangements": Array [
                  Object {
                    "bed": Object {
                      "name": "king",
                    },
                    "people": Array [
                      Object {
                        "gender": "female",
                        "howManyOthersCanShareBed": Object {
                          "king": 2,
                          "queen": 1,
                          "twin": 0,
                        },
                        "name": "Caroline",
                      },
                      Object {
                        "gender": "female",
                        "howManyOthersCanShareBed": Object {
                          "king": 2,
                          "queen": 1,
                          "twin": 0,
                        },
                        "name": "Bethany",
                      },
                    ],
                  },
                ],
              },
              Object {
                "bedArrangements": Array [
                  Object {
                    "bed": Object {
                      "name": "twin",
                    },
                    "people": Array [
                      Object {
                        "gender": "male",
                        "howManyOthersCanShareBed": Object {
                          "king": 2,
                          "queen": 1,
                          "twin": 0,
                        },
                        "name": "Kwinton",
                      },
                    ],
                  },
                  Object {
                    "bed": Object {
                      "name": "twin",
                    },
                    "people": Array [
                      Object {
                        "gender": "male",
                        "howManyOthersCanShareBed": Object {
                          "king": 2,
                          "queen": 1,
                          "twin": 0,
                        },
                        "name": "Isaac",
                      },
                    ],
                  },
                  Object {
                    "bed": Object {
                      "name": "twin",
                    },
                    "people": Array [
                      Object {
                        "gender": "male",
                        "howManyOthersCanShareBed": Object {
                          "king": 2,
                          "queen": 1,
                          "twin": 0,
                        },
                        "name": "Deon",
                      },
                    ],
                  },
                  Object {
                    "bed": Object {
                      "name": "twin",
                    },
                    "people": Array [
                      Object {
                        "gender": "male",
                        "howManyOthersCanShareBed": Object {
                          "king": 2,
                          "queen": 1,
                          "twin": 0,
                        },
                        "name": "Larry",
                      },
                    ],
                  },
                  Object {
                    "bed": Object {
                      "name": "twin",
                    },
                    "people": Array [
                      Object {
                        "gender": "male",
                        "howManyOthersCanShareBed": Object {
                          "king": 2,
                          "queen": 1,
                          "twin": 0,
                        },
                        "name": "Noodle",
                      },
                    ],
                  },
                  Object {
                    "bed": Object {
                      "name": "twin",
                    },
                    "people": Array [],
                  },
                ],
              },
            ],
          },
        ],
      },
    ],
    "listing": Object {
      "buildings": Array [
        Object {
          "floors": Array [
            Object {
              "rooms": Array [
                Object {
                  "beds": Array [
                    Object {
                      "name": "king",
                    },
                    Object {
                      "name": "twin",
                    },
                    Object {
                      "name": "twin",
                    },
                  ],
                },
                Object {
                  "beds": Array [
                    Object {
                      "name": "queen",
                    },
                    Object {
                      "name": "queen",
                    },
                  ],
                },
              ],
            },
            Object {
              "rooms": Array [
                Object {
                  "beds": Array [
                    Object {
                      "name": "king",
                    },
                    Object {
                      "name": "twin",
                    },
                    Object {
                      "name": "twin",
                    },
                  ],
                },
              ],
            },
            Object {
              "rooms": Array [
                Object {
                  "beds": Array [
                    Object {
                      "name": "king",
                    },
                    Object {
                      "name": "twin",
                    },
                  ],
                },
                Object {
                  "beds": Array [
                    Object {
                      "name": "king",
                    },
                    Object {
                      "name": "twin",
                    },
                  ],
                },
                Object {
                  "beds": Array [
                    Object {
                      "name": "king",
                    },
                  ],
                },
                Object {
                  "beds": Array [
                    Object {
                      "name": "twin",
                    },
                    Object {
                      "name": "twin",
                    },
                    Object {
                      "name": "twin",
                    },
                    Object {
                      "name": "twin",
                    },
                    Object {
                      "name": "twin",
                    },
                    Object {
                      "name": "twin",
                    },
                  ],
                },
              ],
            },
          ],
        },
      ],
      "price": 0,
    },
  },
]
`;
